pkg/virtualnode/enqueue_requests_for_backendgroup_events.go

```diff
@@ -0,0 +1,68 @@
+package virtualnode
+
+import (
+	"context"
+	appmesh "github.com/aws/aws-app-mesh-controller-for-k8s/apis/appmesh/v1beta2"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/k8s"
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/util/workqueue"
+	"reflect"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+)
+
+func NewEnqueueRequestsForBackendGroupEvents(client client.Client, log logr.Logger) *enqueueRequestsForBackendGroupEvents {
+	return &enqueueRequestsForBackendGroupEvents{
+		k8sClient: client,
+		log:       log,
+	}
+}
+
+var _ handler.EventHandler = (*enqueueRequestsForBackendGroupEvents)(nil)
+
+type enqueueRequestsForBackendGroupEvents struct {
+	k8sClient client.Client
+	log       logr.Logger
+}
+
+// Create is called in response to a create event
+func (h *enqueueRequestsForBackendGroupEvents) Create(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
+	h.enqueueVirtualNodesForMesh(context.Background(), queue, e.Object.(*appmesh.BackendGroup).Spec.MeshRef)
+}
+
+// Update is called in response to an update event
+func (h *enqueueRequestsForBackendGroupEvents) Update(e event.UpdateEvent, queue workqueue.RateLimitingInterface) {
+	bgOld := e.ObjectOld.(*appmesh.BackendGroup)
+	bgNew := e.ObjectNew.(*appmesh.BackendGroup)
+	if !reflect.DeepEqual(bgOld.Spec.VirtualServices, bgNew.Spec.VirtualServices) {
+		h.enqueueVirtualNodesForMesh(context.Background(), queue, bgNew.Spec.MeshRef)
+	}
+}
+
+// Delete is called in response to a delete event
+func (h *enqueueRequestsForBackendGroupEvents) Delete(e event.DeleteEvent, queue workqueue.RateLimitingInterface) {
+	// no-op
```

Comment:
I agree in terms of the k8s model, though this isn't unique to backend groups as currently virtual nodes don't watch for virtual service deletion.

Reply:
I think the impact of this might be pretty different in that for backend groups you end up with VN instances that still have the previously associated services as backends. In the case of a VS being deleted I don't think the Envoy will get that service as a backend regardless of the fact that it's still configured on the VN. I'm open to whether that's really a concerning difference though. Thoughts?

Reply:
Currently if a VN has a VS reference that can't be found, the reconcile attempt will error out and get re-queued. I followed that same pattern with backend groups. To get rid of backends from deleted backend groups, we'd need to be able to reconcile the VN while ignoring the deleted backend groups. My concern here is this leaves room to miss backend groups if the VN can't find them for some reason.